Labelling workload cluster's apps (#53)
giantswarm/rfc: labelling-wc-apps/README.md

# Labelling workload cluster's apps

The purpose of this RFC is to propose a labelling scheme to safely identify a workload cluster's defining apps.

This is required for:

- deleting a workload cluster on CAPI;

> - displaying a workload cluster's defining apps in Happa.

Reviewer: ```suggestion
- displaying a workload cluster's defining apps in a client like the web UI (happa).
```

## Deleting a workload cluster on CAPI

> To delete a workload cluster on CAPI we need to identify corresponding cluster app.

Reviewer: ```suggestion
To delete a workload cluster on CAPI, we need to identify the corresponding cluster app.
```

### Current state

On CAPI we are relying on the resource names (in the org namespace) for cluster app selection which is not recommended. There is no reference between cluster CR and cluster app CR - unlike other cluster's app resources, cluster app CR doesn't have `giantswarm.io/cluster` label. It cannot be set at the moment, since it will break cluster deletion.

### Proposal

> Require `giantswarm.io/cluster` to be present in all app resources related to a workload cluster, including cluster app.

Reviewer: ```suggestion
Require the `giantswarm.io/cluster` label to be present in all App resources related to a workload cluster, including the cluster app.
```

Introduce a new `giantswarm.io/cluster-app` label to safly mark cluster app CRs.

Refactor places where it's assumed that cluster app doesn't have `giantswarm.io/cluster` label to rely on a new `giantswarm.io/cluster-app` label instead.

## Displaying a workload cluster's defining apps in Happa

In Happa, we display workload cluster's apps in two separate lists - apps installed by a user and default (preinstalled) apps. To achieve that, we need:

> - fetch all App CRs related to a workload cluster. On CAPI this list should not contain cluster app App CR.

Reviewer: Why should this list not include the cluster app?

Author: I should rephrase it. It's more like 
> Currently we don't display cluster app in the UI, so we should either fetch all the apps except cluster app, or we need a way to filter it out later.

Reviewer: Now I get it. Yep, would be good to clarify that.

- split the App CRs into default apps and user installed apps.

### Current state

There is a `giantswarm.io/cluster` label that potentially can be used to address App CRs related to a workload cluster, but:

- it's missing in some App CRs on vintage (currently we fetch all App CRs in a cluster namespace);
- it's missing in a cluster app on CAPI (currently we use `giantswarm.io/cluster` label to select cluster's App CRs in the organization namespace).

There is no suitable way to distinguish apps installed by a user from default apps.

### Proposal

#### For cluster's apps fetching:

On vintage, let's make `giantswarm.io/cluster` label to be required for all app resources related to a workload cluster.

On CAPI, proposal from the "Deleting a workload cluster on CAPI" will also suit this use case. We will fetch all cluster's app resources relying on the `giantswarm.io/cluster` label and the `giantswarm.io/cluster-app` label can be used to filter out cluster app from a list.

#### For splitting cluster's apps into two lists:

> We could eather introduce a new label to mark default apps bundle e.g. `giantswarm.io/installed-by-default`, or we can set `giantswarm.io/managed-by` label of a default apps bundle to something specific (right now it's not set).

Reviewer: ```suggestion
We could either introduce a new label to mark default apps bundle, e.g. `giantswarm.io/installed-by-default`, or we can set the `giantswarm.io/managed-by` label of a default apps bundle to something specific (right now it's not set).
```

> Require that `giantswarm.io/managed-by` label is correctly set for app bundle's child apps. It should point to a parent app.

Reviewer: ```suggestion
Require that `giantswarm.io/managed-by` label is correctly set for app bundle's child apps. It should point to the parent app.
```

# Labelling workload cluster's apps

The purpose of this RFC is to propose a labelling scheme to safely identify a workload cluster's defining apps.

This is required for:

- deleting a workload cluster on CAPI;
- displaying a workload cluster's defining apps in a client like the web UI (happa).

## Deleting a workload cluster on CAPI

To delete a workload cluster on CAPI, we need to identify the corresponding cluster app.

### Current state

On CAPI we are relying on the resource names (in the org namespace) for cluster app selection which is not recommended. There is no reference between cluster CR and cluster app CR - unlike other cluster's app resources, cluster app CR doesn't have `giantswarm.io/cluster` label. It cannot be set at the moment, since it will break cluster deletion.

### Proposal

> Require the `giantswarm.io/cluster` label to be present in all App resources related to a workload cluster, including the cluster app.

Reviewer 2: Just to add some context, I believe that we don't add the label to the cluster-app because `cluster-apps-operator` adds a finalizer to the Cluster CR once it starts reconciling a cluster, which is removed from the Cluster CR once all the apps labeled `giantswarm.io/cluster` are deleted. But the cluster-app won't be removed until the Cluster CR is removed. And that won't happen until the finalizer is removed. And that won't happen until the cluster-app is removed. So we have a chicken - egg kind of problem.

How do you propose to rewrite the deletion logic so that it works after adding the `giantswarm.io/cluster` label?

Author: I think we can be more explicit about not deleting the cluster-app. At the moment it kinda happens as a side effect, it's assumed that you have a knowledge that the cluster-app doesn't have a `giantswarm.io/cluster` label. Maybe instead of selecting the apps labeled with `giantswarm.io/cluster` label, we can select the apps labeled with `giantswarm.io/cluster` and that don't have `application.giantswarm.io/cluster-app` label which I suggest to add.
